### DISCUSSION TITLE
componentsjs-generator outputs the error "Could not load class or interface or other type Omit from ..."

### DISCUSSION THREAD

**Discussion by ksaito-hiu** [LinkedSoftwareDependencies/Components.js#139] [Q&A] [answered]
Hello,

I'm interested in CommunitySolidServer and am learning TypeScript and Components.js.

I want to test AccountLoginStorage in CSS to add a new login method.
To do this, I added a constructor to the HelloWorld class in the hello-world-component repository of the CSS tutorial so that it could receive AccountLoginStorage.

hello-world-component:src/HelloWorld.ts

```
import { getLoggerFor, Initializer } from '@solid/community-server';
import { AccountLoginStorage } from '@solid/community-server';

export const HELLO_STORAGE_TYPE = 'hello';
export const HELLO_STORAGE_DESCRIPTION = {
  hello: 'string',
} as const;

/**
 * An {@link Initializer} that logs an informative message during startup.
 */
export class HelloWorld extends Initializer {
  protected readonly logger = getLoggerFor(this);

  private readonly storage: AccountLoginStorage<{ [HELLO_STORAGE_TYPE]: typeof HELLO_STORAGE_DESCRIPTION }>;

  // Wrong typings to prevent Components.js typing issues
  public constructor(storage: AccountLoginStorage<Record<string, never>>) {
    super();
    this.storage = storage as unknown as typeof this.storage;
  }

  public async handle(input: void): Promise<void> {
    this.logger.info('HELLO WORLD!');
  }
}
```

hello-world-component:config/hello-world.json

```
{
  "@context": "https://linkedsoftwaredependencies.org/bundles/npm/hello-world-module/^7.0.0/components/context.jsonld",
  "@graph": [
    {
      "comment": "A complete Hello World instance",
      "@type": "HelloWorld",
      "@id": "urn:my-hello-world:core:HelloWorld",
      "storage": "urn:solid-server:default:AccountStorage"
    }
  ]
}
```

The above program was based on BasePasswordStore.ts.
As a result, componentsjs-generator now outputs the following error.

```
Could not load class or interface or other type Omit from /home/ksaitoSubHome/src/git/hello-world-component/node_modules/@solid/community-server/dist/identity/interaction/account/util/LoginStorage
```

Maybe it's because I don't have enough knowledge about TypeScript types, but I can't figure out what's wrong.

You can see the complete diff in my github repository, which I created by forking hello-world-component.

https://github.com/CommunitySolidServer/hello-world-component/compare/main...ksaito-hiu:hello-world-component:main

I would be very grateful if you could give me some advice.

**Comment by joachimvh** [comment, marked as answer]
The problem is a consequence of this issue: https://github.com/LinkedSoftwareDependencies/Components-Generator.js/issues/80. The Components.js generator has problems with native Node.js builtins, such as `Omit`. You can tell it to ignore such classes by having a `.componentsignore` file and adding `-i .componentsignore` to your `components-generator` call and having a JSON array where each entry is a string corresponding to a class/interface the Components.js generator should ignore. So in this case you would `[ "Omit" ]` in that file. The hello world example already has such a file where you can just fill this in.

Although your class does not use `Omit`, you are depending on a class that does, causing this issue. You can see that in the CSS repo this (and many others) in its `.componentsignore` file: https://github.com/CommunitySolidServer/CommunitySolidServer/blob/486241f3d496a972518474f0be74b79a11e2db2b/.componentsignore#L28

**Comment by ksaito-hiu** [comment]
Thank you very much. I could fix the bug by adding "Omit" entry in .componentsignore!

I still don't understand why the "Omit" type is involved, but thanks to joachimvh's advice I think I can go ahead and understand it better.

In case others may refer this Discussion, I would like to point out that the hello-world.json I presented above contained another bug.
The value of the "storage" argument is an object, so I have to write { "@id": "urn:solid-server:default:AccountStorage" }. The whole thing will look like this:

```
{
  "@context": "https://linkedsoftwaredependencies.org/bundles/npm/hello-world-module/^7.0.0/components/context.jsonld",
  "@graph": [
    {
      "comment": "A complete Hello World instance",
      "@type": "HelloWorld",
      "@id": "urn:my-hello-world:core:HelloWorld",
      "storage": { "@id": "urn:solid-server:default:AccountStorage" }
    }
  ]
}
```

Thank you very much for your accurate advice.

**Reply by joachimvh** [reply]
> I still don't understand why the "Omit" type is involved

Because the `AccountLoginStorage` class you added as a parameter has it somewhere in its typings. (Specifically [here](https://github.com/CommunitySolidServer/CommunitySolidServer/blob/def0b5c732de38d76d90d05332a5380717652f8c/src/identity/interaction/account/util/LoginStorage.ts#L17))

**Reply by ksaito-hiu** [reply]
Oh! AccountLoginStorage is a type alias of LoginStorage, and LoginStorage extends Omit. I understand. Thanks again.
